glide-core/src/client/value_conversion.rs

```diff
@@ -673,6 +674,237 @@ pub(crate) fn convert_to_expected_type(
+                    5# "pending" =>
+                       1) 1) "1719710679916-0"
+                          2) (integer) 1719710707260
+                          3) (integer) 1
+                       2) 1) "1719710688676-0"
+                          2) (integer) 1719710718373
+                          3) (integer) 1
+        
+        Another RESP3 example on an empty stream
+
+        1# "length" => (integer) 0
+        2# "radix-tree-keys" => (integer) 0
+        3# "radix-tree-nodes" => (integer) 1
+        4# "last-generated-id" => "0-1"
+        5# "max-deleted-entry-id" => "0-1"
+        6# "entries-added" => (integer) 1
+        7# "recorded-first-entry-id" => "0-0"
+        8# "entries" => (empty array)
+        9# "groups" => (empty array)
+
+        We want to convert the RESP2 format to RESP3, so we need to:
+        - convert any consumer in the consumer array to a map, if there are any consumers
+        - convert any group in the group array to a map, if there are any groups
+        - convert the root of the response into a map
+        */
+        ExpectedReturnType::XInfoStreamFullReturnType => match value {
+            Value::Map(_) => Ok(value),  // Response is already in RESP3 format - no conversion needed
+            Value::Array(mut array) => {
+                // Response is in RESP2 format. We need to convert to RESP3 format.
+                let groups_key = Value::SimpleString("groups".into());
+                let opt_groups_key_index = array
+                    .iter()
+                    .position(
+                        |key| {
+                            let res = convert_to_expected_type(key.clone(), Some(ExpectedReturnType::SimpleString));
+                            match res {
+                                Ok(converted_key) => {
+                                    converted_key == groups_key
+                                },
+                                Err(_) => {
+                                    false
+                                }
+                            }
+                        }
+                    );
+
+                let Some(groups_key_index) = opt_groups_key_index else {
+                    return Err((ErrorKind::TypeError, "No groups key found").into());
+                };
+
+                let groups_value_index = groups_key_index + 1;
+                if array.get(groups_value_index).is_none() {
+                    return Err((ErrorKind::TypeError, "No groups value found.").into());
+                }
+
+                let Value::Array(groups) = array[groups_value_index].clone() else {
+                    return Err((ErrorKind::TypeError, "Incorrect value type received. Wanted an Array.").into());
+                };
+
+                if groups.is_empty() {
+                    let converted_response = convert_to_expected_type(Value::Array(array), Some(ExpectedReturnType::Map {
+                        key_type: &Some(ExpectedReturnType::BulkString),
```

Comment:
This is the scenario where "groups" is empty, ie there are no groups. Usually the nested maps are inside of groups (each group needs to be converted to a map, and each consumer inside that group map needs to be converted to a map). But since groups is empty there are no nested maps that we need to convert

Reply:
Oh I see
So you just flat the recursion into a nested loops.
Maybe we need to rework it in the CompletableFuture